replacement_mode_modeling/02_run_trip_level_models.py

```diff
@@ -0,0 +1,491 @@
+from enum import Enum
+import random
+import warnings
+import argparse
+from pathlib import Path
+from collections import Counter
+
+# Math and graphing.
+import pandas as pd
+import numpy as np
+import seaborn as sns
+import matplotlib.pyplot as plt
+
+# sklearn imports.
+from sklearn.model_selection import train_test_split
+from sklearn.preprocessing import StandardScaler
+from sklearn.linear_model import LinearRegression
+from sklearn.metrics import f1_score, r2_score, ConfusionMatrixDisplay
+from scipy.special import kl_div
+from sklearn.metrics import classification_report
+from sklearn.model_selection import GridSearchCV, StratifiedGroupKFold
+from pprint import pprint
+from sklearn.inspection import permutation_importance
+from time import perf_counter
+from sklearn.ensemble import RandomForestClassifier
+
+warnings.simplefilter(action='ignore', category=Warning)
+
+# Global experiment flags and variables.
+SEED = 13210
+TARGETS = ['p_micro', 'no_trip', 's_car', 'transit', 'car', 's_micro', 'ridehail', 'walk', 'unknown']
+MAP = {ix+1:t for ix, t in enumerate(TARGETS)}
+
+CV = False
+
+# Set the Numpy seed too.
+random.seed(SEED)
+np.random.seed(SEED)
+
+class SPLIT_TYPE(Enum):
+    INTRA_USER = 0
+    INTER_USER = 1
+    TARGET = 2
+    MODE = 3
+    HIDE_USER = 4
+    
+
+class SPLIT(Enum):
+    TRAIN = 0
+    TEST = 1
+
+
+def get_train_test_splits(data: pd.DataFrame, how=SPLIT_TYPE, test_ratio=0.2, shuffle=True):
+    
+    if how == SPLIT_TYPE.INTER_USER:
+
+        X = data.drop(columns=['target'])
+        y = data['target'].values
+        groups = data.user_id.values
+
+        # n_splits determines split size. So n=5, is 20% for each split, which is what we want.
+        splitter = StratifiedGroupKFold(n_splits=5, shuffle=shuffle, random_state=SEED)
+        # splitter = GroupKFold(n_splits=5)
+        
+        for train_index, test_index in splitter.split(X, y, groups):
```

Comment:
We do so create one chunk of train and test sets with 80% and 20% of the stratified data respectively. sklearn's reguar `train_test_split` method unfortunately does not provide a way to create training and testing splits with groups. There may be other ways to achieve this, but this is what I went with for now.


For clarification, StratifiedGroupKFold creates splits by stratifying w.r.t. the number of trips per user. The algorithm then attempts to preserve the percentage of groups in each split. I purposely mention `n=5 splits`, because that is the information that the algorithm uses to determine the percentage of data in each split. 


n=5 => 20% of test data. Therefore we create one split, get the training and testing data, and break the iteration.